### PR TITLE
Add middleware interface

### PR DESCRIPTION
**Pull request by ameech**
Implementing middleware similar to https://github.com/radarphp/Radar.Project/blob/1.x/docs/middleware.md

### CODE REVIEW

**Comment by shadowhand** [comment]
:+1:

File: src/MiddlewareInterface.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+namespace Spark\Adr;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+interface MiddlewareInterface
+{
+
+    /**
+     * Processes requests and responses
```
Comment (shadowhand) [inline src/MiddlewareInterface.php]: I would update this comment to be a bit more descriptive about what middleware does: modify the request/response and pass it on to the next middleware.

**Comment by ameech** [comment]
:dancer: There we go.

File: src/MiddlewareInterface.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+     *
+     * @param ServerRequestInterface $request
+     * @param ResponseInterface $response
+     * @param callable $next
+     * @return ResponseInterface
+     */
+    public function __invoke(
+        ServerRequestInterface $request,
+        ResponseInterface $response,
+        callable $next
```
Comment (shadowhand) [inline src/MiddlewareInterface.php]: Just a thought on this... I _think_ the final call for middleware would be run like this:

```php
// in Application::handle
$last = function ($request, $response) {
    return $response;
};
foreach ($middlewares as $i => $middle) {
    $next = isset($middlewares[$i + 1]) ? $middlewares[$i] : $last;
    $response = $middle($request, $response, $next);
}
```

Maybe you already had a plan for this, I was just trying to work it out in my head.

Comment (shadowhand) [inline src/MiddlewareInterface.php]: Wait nevermind, that doesn't make any sense... got too far ahead of myself.

File: src/MiddlewareInterface.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+interface MiddlewareInterface
+{
+
+    /**
+     * Modify the request/response and pass it to the next middleware
+     *
+     * @param ServerRequestInterface $request
+     * @param ResponseInterface $response
+     * @param callable $next
+     * @return ResponseInterface
```
Comment (dolfelt) [inline src/MiddlewareInterface.php]: Should this return an array with `ServerRequestInterface` and `ResponseInterface`? Based on https://github.com/sparkphp/Spark/pull/11/files#diff-c9248b3167fc44af085b81db2e292837R255 it should.

Technically the `ServerRequestInterface` is immutable, but perhaps there are use cases for modifying it.

Comment (ameech) [inline src/MiddlewareInterface.php]: It should, that was a mistake on my part.

Comment (dolfelt) [inline src/MiddlewareInterface.php]: :+1: Awesome.

**Pull request merged**